velox/functions/sparksql/aggregates/AverageAggregate.cpp

```diff
@@ -74,6 +74,319 @@ class AverageAggregate
+    auto countVector = rowVector->childAt(1)->asFlatVector<int64_t>();
+    rowVector->resize(numGroups);
+    sumVector->resize(numGroups);
+    countVector->resize(numGroups);
+    rowVector->clearAllNulls();
+
+    int64_t* rawCounts = countVector->mutableRawValues();
+    int128_t* rawSums = sumVector->mutableRawValues();
+    for (auto i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      auto* accumulator = this->decimalAccumulator(group);
+      rawCounts[i] = accumulator->count;
+      rawSums[i] = accumulator->sum;
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto vector = (*result)->as<FlatVector<TResultType>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+    uint64_t* rawNulls = this->getRawNulls(vector);
+
+    TResultType* rawValues = vector->mutableRawValues();
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      auto accumulator = this->decimalAccumulator(group);
+      if (accumulator->count == 0) {
+        // In Spark, if all inputs are null, count will be 0,
+        // and the result of final avg will be null.
+        vector->setNull(i, true);
+      } else {
+        this->clearNull(rawNulls, i);
+        try {
+          rawValues[i] = computeAvg(accumulator);
+        } catch (const VeloxException& err) {
+          if (err.message().find("overflow") != std::string::npos ||
+              err.message().find("is not in the range of Decimal Type") !=
+                  std::string::npos) {
+            // find overflow or out of long decimal range in computation
+            vector->setNull(i, true);
+          } else {
+            VELOX_FAIL("compute average failed");
+          }
+        }
+      }
+    }
+  }
+
+  TResultType computeAvg(LongDecimalWithOverflowState* accumulator) {
+    int128_t sum = accumulator->sum;
+    if ((accumulator->overflow == 1 && accumulator->sum < 0) ||
+        (accumulator->overflow == -1 && accumulator->sum > 0)) {
+      sum = static_cast<int128_t>(
+          DecimalUtil::kOverflowMultiplier * accumulator->overflow +
+          accumulator->sum);
+    } else {
+      VELOX_CHECK(
+          accumulator->overflow == 0,
+          "overflow: decimal avg struct overflow not eq 0");
+    }
+
+    auto [resultPrecision, resultScale] =
+        getDecimalPrecisionScale(*this->resultType().get());
+    // Spark use DECIMAL(20,0) to represent long value
+    int countPrecision = 20;
+    int countScale = 0;
+    uint8_t sumPrecision = resultPrecision - 4 + 10;
+    uint8_t sumScale = resultScale - 4;
+    auto [avgPrecision, avgScale] = computeResultPrecisionScale(
+        sumPrecision, sumScale, countPrecision, countScale);
+    auto sumRescale = computeRescaleFactor(sumScale, countScale, avgScale);
+    auto countDecimal = accumulator->count;
+    int128_t avg = 0;
+
+    DecimalUtil::divideWithRoundUp<int128_t, int128_t, int128_t>(
```

Comment:
Maybe you can wait for this to add overflow flag. https://github.com/facebookincubator/velox/pull/4613/files#diff-4b21d26191cbbdb9bc90cafd83aed144a4483b1c594f224cf3bbe67f401f5f83R86